PWAWidgets/README.md

```diff
@@ -0,0 +1,975 @@
+1. Append <var>instance</var> to <var>widget["instances"]</var>.
+
+<p id="install">Here is the flow for install:</p>
+
+![](media/install.gif)
+
+1. A "WidgetInstall" signal is received by the User Agent, the placeholder instance is created, and the event is passed along to the Service Worker.
+2. The Service Worker
+    a. captures the Widget Instance `id` from the `widget` property,
+    b. looks up the Widget via `widgets.getByInstance()`, and
+    c. makes a `Request` for its `data` endpoint.
+3. The Service Worker then combines the `Response` with the Widget definition and passes that along to the [Widget Service](#dfn-widget-service) via the `updateInstance()` method.
+
+### WidgetUninstall
+
+Required `WidgetEvent` data:
+
+* `host`
+* `instance`
+* `tag`
+
+<p id="uninstall">The "uninstall" process is similar:</p>
+
+![](media/uninstall.gif)
+
+1. The "WidgetUninstall" signal is received by the User Agent and is passed to the Service Worker.
+1. The Service Worker runs any necessary cleanup steps (such as un-registering a Periodic Sync if the widget is no longer in use).
+1. The Service Worker calls `removeInstance()` to complete the removal process.
+
+Note: When a PWA is uninstalled, its widgets must also be uninstalled. In this event, the User Agent must prompt the [Widget Service](#dfn-widget-service) to remove all associated widgets. If the UA purges all site data and the Service Worker during this process, no further steps are necessary. However, if the UA does not purge all data, it must issue uninstall events for each Widget Instance so that the Service Worker may unregister related Periodic Syncs and perform any additional cleanup.
+
+### WidgetSave
+
+Required `WidgetEvent` data:
+
+* `host`
+* `instance`
+* `tag`
+* `data`
+
+The "WidgetSave" process works like this:
+
+1. The "WidgetSave" signal is received by the User Agent.
+2. Internally, the `WidgetInstance` matching the `widget.id` value is examined to see if
+   a. it has settings and
+   b. its `settings` object matches the inbound `data`.
+3. If it has settings and the two do not match, the new `data` is saved to `settings` in the `WidgetInstance` and the "WidgetSave" event issued to the Service Worker.
+4. The Service Worker receives the event and can react by issuing a request for new data, based on the updated settings values.
+
+### WidgetResume
+
+Many [Widget Hosts](#dfn-widget-host) will suspend the rendering surface when it is not in use (to conserve resources). In order to ensure Widgets are refreshed when the rendering surface is presented, the [Widget Host](#dfn-widget-host) will issue a "WidgetResume" event.
+
+Required `WidgetEvent` data:
+
+* `host`
+
+Using this event, it is expected that the Service Worker will enumerate the Widget Instances associated with the `host` and Fetch new data for each.
+
+![](media/resume.gif)
```

Comment:
And here

Reply:
The text below fully describes the image, which is why I didn’t include it.